### ISSUE TITLE
Document migration steps for moving between database backends (e.g. SQLite to MySQL/PGSQL)

### ISSUE THREAD

**Issue by zertrin** [wallabag/wallabag#4126]
(this is a follow-up from the discussion https://github.com/wallabag/wallabag/issues/2766#issuecomment-535070032 )

In the documentation, there are 2 cases summarily documented:

- [Backup](https://doc.wallabag.org/en/admin/backup.html): which briefly describe backing up the config file, database and local images
- [Import](https://doc.wallabag.org/en/user/import/wallabagv2.html): which basically tells us to export all article as JSON and reimport them.

If a user wants to migrate between database backends, the first approach won't help, because we would be simply restoring to the same backend.

That leaves us with the second approach, namely exporting all articles as JSON and re-importing it in a fresh instance which has been initialized with the new database backend. However, this has some limitations, because the JSON export file does not really contain 100% of the information that is stored in the old wallabag instance. For example, tagging rules or local images (and probably more?).

This issue here is about documenting two independent but complementary TODOs:

- [ ] what **is** and **is not** included in the JSON export
- [ ] what are the recommended steps to migrate from a backend (say SQLite) to another backend (MySQL or PostgreSQL) _**with the least amount of information loss during the migration**_.

I am ready to help and propose a PR for the documentation after testing it myself (I am in the situation that I want to migrate from SQLite). 

However I need the help of the wallabag developers to tell me how to minimize the information loss due to the migration.

**Comment by zertrin** [comment]
Ok so today I have attempted the migration from SQLite to MariaDB. Time for some feedback. 😄 

I have git cloned wallabag into a new folder. Then I ran `make install` and followed the instructions to connect to mysql.

I exported all entries as JSON from the web interface on the old install. I got a 12 MB json file.

After that, I switched the VHosts so the website points to the new install. I attempted an import from the web interface, but it failed quickly saying the file is too big (but without saying by how much... that's a UX issue that could be addressed separately).

So I proceeded by importing from the command line (with redis used for asynchronus import). 

It took a while and there was *a lot* of errors about importing images. I suppose that if the image is not available online anymore, then there are problems... **This is one of the things that simply exporting and importing the JSON export will not do for you and that I would like to be able to migrate easily from the old install**. Some of the articles in wallabag are old and not available online anymore...

~I had a quick look into the `new-wallabag-install/web/assets/images` directory structure, but it seems the hashes used there are not deterministic. During the import, I saw some errors for a path, and when I tried to find it in the old `old-wallabag-install/web/assets/images` directory, it is not there... I am in the opinion that such things should be deterministic, to allow migration.~

EDIT: I was apparently mistaken, it's just that the ones I happened to look at were not available for some reason.

In the end, I decided to try the brute-force approach and manually merged the `web/assets/images` folders like this:

    rsync -avh --progress --ignore-existing /srv/wallabag2{-old,-new}/web/assets/images/

I cannot check all articles manually, but after looking at a few of them, it might (not sure 100%) have done the trick.

Another "surprise" after the import: **the entries were reimported in reverse order**. So the most recent entry has ID 1 while the oldest one has ID 413. 

This surprises me a second time, because I am supposed to have 473 entries in total!? So how come the max ID is 413? Well after looking in the database, I do have entries up to ID 473, but this one is not the oldest entry somehow? Something is not very logical here.

I reconfigured the wallabager extension for FF and the Android App with new API credentials.

I'll try using the migrated version a while to see if there are any remaining problems I haven't noticed yet and report here if needed.

**Comment by papylhomme** [comment]
Hello, I'm interested by some feedback as I'm planning on migrating from MariaDB to Postgres. Did you encounter other problems and reverted or are you still using your new instance ?

**Comment by zertrin** [comment]
> Hello, I'm interested by some feedback as I'm planning on migrating from MariaDB to Postgres. Did you encounter other problems and reverted or are you still using your new instance ?

I've simply used the new instance since then. 

I still consider that the documentation is lacking with regards to migrating between backends, and that the export/import approach has significant drawbacks.

**Comment by papylhomme** [comment]
I totally agree with you, that's the main reason why I hadn't attempted it. Thanks for the details BTW !
Trying it now, but the postgres import is failing with not-null constraints violation (#4547)

**Comment by bigretromike** [comment]
Same here, exported to json (33MB) and while importing `The uploaded file was too large. Please try to upload a smaller file.`

**Comment by johnchristopher** [comment]
Has anyone make any progress towards a step-by-step procedure to move from sqlite backend to mysql ?

I have postponed it for 3 years but I am still running 2.4.1. From what I read JSON export doesn't migrate everything and there are no export script for an SQL dump compatible with Mysql.

JSON export runs out of memory on my installation so my only option is to install a second wallabag instance and consider that sqlite installation of wallabag simply can't be migrated. Correct ?

**Comment by jivanyatra** [comment]
I'm thinking about possibly making a basic tool to help with this, as I'm still using sqlite and need to migrate to something else (probably postgres).

I want to make clear: I'm not promising anything, but just thinking and hopefully engaging in discussion right now. Time is in limited supply, after all.

It seems like the primary issues are surrounding exporting and importing via JSON. However, I've done many migrations and restores manually just fine, and I've been backing up the images and db to do it. The settings must be present somewhere in the db or the config folder so as to allow this.

I'm wondering if it makes sense to write something that will map db schemas from sqlite to postgres and then run the data migration. We'd get the image locations and IDs and everything to match this way. If it works, it'd serve as a template for others to map to MySQL (which is the suggested and default db these days), or I could do it when I get around to it as a second step.

An alternative that's halfway done for me (because of a personal tool I already made) is to sync the articles via API. A runner that pulls articles from the old instance into a newly configured instance could migrate content more transparently to the end user. The article IDs would change because those can't be set, and IDK about image data. But, it'd save me time and I'd only have to map the settings tables schemas...

I've written a script that grabs articles and searches them for linked tags via regex, as many sites embed links to their own tags. I have a runner for the API already because of this. As an aside, I'm patiently waiting for webhooks so I can use them instead of running cron and keeping track myself.

Thanks to the dev team and contributors for all the work they've done over the years to make this a real and viable alternative to the services out there harvesting our data.

**Comment by yguedidi** [comment]
I think the best option is to migrate thanks to SQL dumps of the database.
The database schema is the same whatever the database used.
You'll need the command line tools of each database and have them configured to access their respective database.

For example, for SQLite you can dump all the data like this (from [here](https://stackoverflow.com/questions/75675/how-to-dump-the-data-of-some-sqlite3-tables#answer-7526055))

> You can do this getting difference of .schema and .dump commands. for example with grep:
> 
> ```
> sqlite3 some.db .schema > schema.sql
> sqlite3 some.db .dump > dump.sql
> grep -vx -f schema.sql dump.sql > data.sql
> ```
> data.sql file will contain only data without schema, something like this:
> 
> ```
> BEGIN TRANSACTION;
> INSERT INTO "table1" VALUES ...;
> ...
> INSERT INTO "table2" VALUES ...;
> ...
> COMMIT;
> ```

Then, change wallabag settings to used the new empty PostgreSQL database, run the `wallabag:install` command like a new installation to create the database schema, but do not create a default user during the process.

Once done, you could execute the data dump file in PostgreSQL like this (from [here](https://www.postgresql.org/docs/current/backup-dump.html#BACKUP-DUMP-RESTORE)):

> `psql dbname < dumpfile`

**Note**: in theory it should work, not tested on my side.

**Comment by jivanyatra** [comment]
I'm finally coming back around to this. I've changed away from using an installed version to using docker for my deployments. Do I need a custom image which stops after the `wallabag:install` command and doesn't go on to create a new user to create the schema? Or is there a convenient place to grab the schema from?

**Comment by jivanyatra** [comment]
Okay, documenting what I've got so far:

Got a `docker-compose.yml` file up that has postgres 18.1 (alpine3.23). Once you bring up the wallabag container, you can manually drop into the container and then run the manual installation command, which will create the schemas in the database:

```
docker exec -it container_name sh
bin/console wallabag:install --env=prod --no-interaction

# Since I'm using postgres and not mysql, I didn't need to run this, but I'll provide it here for others following along:
bin/console doctrine:migrations:migrate --env=prod --no-interaction
```

This gets the DB into a workable shape where you can log in successfully to wallabag.

After trying `psql dbname < dumpfile` (and the like), it fails quite a bit. Looks like the issue is that the syntax for sqlite3 != postgres, so having to adjust that in bulk is a pain, but it should work fine after that in theory? I'm going to break up my data-dump table by table and check syntax.

Perhaps this is why mysql/mariadb was recommended? I might try that first and see how the importing goes, if it has less friction, etc.

**Comment by jivanyatra** [comment]
**TL;DR:** [I figured out how to do it.](https://github.com/jivanyatra/wallabag_database_migration)

## Details:

I ended up sticking with postgres once I found [pgloader](https://pgloader.readthedocs.io/en/latest/quickstart.html).

I tried a lot of stuff, resetting the db until I found what worked, which is:
```
pgloader --with "prefetch rows = 100" --dynamic-space-size 1000 wallabag.sqlite postgres://username:password@localhost:5432/wallabag
```
I needed to lower the prefetch row count and specify 1GB of RAM allowance because of the specs on my virtual machine. If you see heap errors, those two options made it work.

I still got an error:
```
2025-12-22T06:36:24.622980Z ERROR PostgreSQL Database error 42P16: multiple primary keys for table "wallabag_internal_setting" are not allowed
QUERY: ALTER TABLE wallabag_internal_setting ADD PRIMARY KEY USING INDEX idx_16773_sqlite_autoindex_wallabag_internal_setting_1;
2025-12-22T06:36:24.744980Z LOG report summary reset
```

I manually checked and the db has the correct values and stuff in place. BUT this clued me into another issue I saw...

If you try to do anything - add a user, save a new entry, create a new tag - you'll encounter 500 errors! After analyzing the above and digging a bit, these 500s come because the db can't add a row with the same id that exists in the db already. That explains why you can't login via API, either - no new entries in the oauth2 tables.

This is because of an issue spotted in [Wallabag Issue # 5502](https://github.com/wallabag/wallabag/issues/5502#issuecomment-1006871510). Basically, either because of the postgres driver or something else, when you create the postgres db with wallabag's installation, the sequences used for the tables' `id`s are created without the `wallabag_` prefix. When we import from sqlite, we're basically adding in the state we had from before, but they're not being used by the postgres driver.

I took this opportunity to do some cleanup. I reset the values of the sequences to the highest `id` in their respective tables, so all new entries would appear afterwards. Then, I removed the sequences imported from sqlite, did some testing, and everything looks good to go.


```sql
BEGIN;

ALTER TABLE wallabag_annotation
  ALTER COLUMN id SET DEFAULT nextval('annotation_id_seq');
SELECT setval('annotation_id_seq', (SELECT MAX(id) FROM wallabag_annotation));

ALTER TABLE wallabag_config
  ALTER COLUMN id SET DEFAULT nextval('config_id_seq');
SELECT setval('config_id_seq', (SELECT MAX(id) FROM wallabag_config));

ALTER TABLE wallabag_entry
  ALTER COLUMN id SET DEFAULT nextval('entry_id_seq');
SELECT setval('entry_id_seq', (SELECT MAX(id) FROM wallabag_entry));

ALTER TABLE wallabag_ignore_origin_instance_rule
  ALTER COLUMN id SET DEFAULT nextval('ignore_origin_instance_rule_id_seq');
SELECT setval('ignore_origin_instance_rule_id_seq', (SELECT MAX(id) FROM wallabag_ignore_origin_instance_rule));

ALTER TABLE wallabag_ignore_origin_user_rule
  ALTER COLUMN id SET DEFAULT nextval('ignore_origin_user_rule_id_seq');
SELECT setval('ignore_origin_user_rule_id_seq', (SELECT MAX(id) FROM wallabag_ignore_origin_user_rule));

ALTER TABLE wallabag_oauth2_access_tokens
  ALTER COLUMN id SET DEFAULT nextval('oauth2_access_tokens_id_seq');
SELECT setval('oauth2_access_tokens_id_seq', (SELECT MAX(id) FROM wallabag_oauth2_access_tokens));

ALTER TABLE wallabag_oauth2_auth_codes
  ALTER COLUMN id SET DEFAULT nextval('oauth2_auth_codes_id_seq');
SELECT setval('oauth2_auth_codes_id_seq', (SELECT MAX(id) FROM wallabag_oauth2_auth_codes));

ALTER TABLE wallabag_oauth2_clients
  ALTER COLUMN id SET DEFAULT nextval('oauth2_clients_id_seq');
SELECT setval('oauth2_clients_id_seq', (SELECT MAX(id) FROM wallabag_oauth2_clients));

ALTER TABLE wallabag_oauth2_refresh_tokens
  ALTER COLUMN id SET DEFAULT nextval('oauth2_refresh_tokens_id_seq');
SELECT setval('oauth2_refresh_tokens_id_seq', (SELECT MAX(id) FROM wallabag_oauth2_refresh_tokens));

ALTER TABLE wallabag_site_credential
  ALTER COLUMN id SET DEFAULT nextval('site_credential_id_seq');
SELECT setval('site_credential_id_seq', (SELECT MAX(id) FROM wallabag_site_credential));

ALTER TABLE wallabag_tag
  ALTER COLUMN id SET DEFAULT nextval('tag_id_seq');
SELECT setval('tag_id_seq', (SELECT MAX(id) FROM wallabag_tag));

ALTER TABLE wallabag_tagging_rule
  ALTER COLUMN id SET DEFAULT nextval('tagging_rule_id_seq');
SELECT setval('tagging_rule_id_seq', (SELECT MAX(id) FROM wallabag_tagging_rule));

ALTER TABLE wallabag_user
  ALTER COLUMN id SET DEFAULT nextval('user_id_seq');
SELECT setval('user_id_seq', (SELECT MAX(id) FROM wallabag_user));


DROP SEQUENCE wallabag_annotation_id_seq;
DROP SEQUENCE wallabag_config_id_seq;
DROP SEQUENCE wallabag_entry_id_seq;
DROP SEQUENCE wallabag_oauth2_access_tokens_id_seq;
DROP SEQUENCE wallabag_oauth2_auth_codes_id_seq;
DROP SEQUENCE wallabag_oauth2_clients_id_seq;
DROP SEQUENCE wallabag_oauth2_refresh_tokens_id_seq;
DROP SEQUENCE wallabag_tag_id_seq;
DROP SEQUENCE wallabag_tagging_rule_id_seq;
DROP SEQUENCE wallabag_user_id_seq;

COMMIT;
```

I left these three sequences in place. I'm not sure why, but they were created during the `wallabag:install` process, even though they seem vestigial
```
wallabag_ignore_origin_user_rule_id_seq
ignore_origin_instance_rule_id_seq
site_credential_id_seq
```

After all of this, I ran things and everything seems to be working okay. I can add new entries, connect via API/mobile, and I get no 500s. I'm calling this complete - at least for me and my needs.

I've documented all of this [in a repo](https://github.com/jivanyatra/wallabag_database_migration), linked in the TL;DR above. If anyone would like to use any or all of it for official wallabag documentation, please feel free to. I'll drop a permissible license in the repo as well.

**Comment by jivanyatra** [comment]
Welp, spoke too soon of course.

Turns out my above method makes a mess of things. The postgres tables and many of the columns aren't quite right (specifically types) at the end of the process, and may cause issues down the line. I figured this out when I tried to delete an entry with a tag - the `wallabag_entry_tag` table had a missing `ON CASCADE DELETE`.

I redid my entire [instruction set in my repo](https://github.com/jivanyatra/wallabag_database_migration). I wrote a [pg_loader commands file](https://github.com/jivanyatra/wallabag_database_migration/blob/main/wallabag_import_commands_pgloader.load) that casts the data type of the sqlite3 columns  to the destination types in the postgres db.

There's a further mismatch between wallabag's sqlite and postgres tables. I had to rename my sqlite tables to match (though I guess you could figure out a view or something in the pgloader commands file so you don't have to alter it in sqlite?).

```
BEGIN;
ALTER TABLE wallabag_user RENAME COLUMN authCode TO authcode ;
ALTER TABLE wallabag_user RENAME COLUMN emailTwoFactor TO emailtwofactor ;
ALTER TABLE wallabag_user RENAME COLUMN googleAuthenticatorSecret TO googleauthenticatorsecret ;
ALTER TABLE wallabag_user RENAME COLUMN backupCodes TO backupcodes ;
ALTER TABLE wallabag_site_credential RENAME COLUMN createdAt TO createdat ;
COMMIT;
```

Then I did a fresh init of the wallabag database into postgres. Then I used pg_loader to import my commands.load file. Success!

Lastly, we have to reset the sequences via `psql` so new entries get added correctly:

```
BEGIN;
SELECT setval('annotation_id_seq', (SELECT MAX(id) FROM wallabag_annotation));
SELECT setval('config_id_seq', (SELECT MAX(id) FROM wallabag_config));
SELECT setval('entry_id_seq', (SELECT MAX(id) FROM wallabag_entry));
SELECT setval('ignore_origin_instance_rule_id_seq', (SELECT MAX(id) FROM wallabag_ignore_origin_instance_rule));
SELECT setval('ignore_origin_user_rule_id_seq', (SELECT MAX(id) FROM wallabag_ignore_origin_user_rule));
SELECT setval('oauth2_access_tokens_id_seq', (SELECT MAX(id) FROM wallabag_oauth2_access_tokens));
SELECT setval('oauth2_auth_codes_id_seq', (SELECT MAX(id) FROM wallabag_oauth2_auth_codes));
SELECT setval('oauth2_clients_id_seq', (SELECT MAX(id) FROM wallabag_oauth2_clients));
SELECT setval('oauth2_refresh_tokens_id_seq', (SELECT MAX(id) FROM wallabag_oauth2_refresh_tokens));
SELECT setval('site_credential_id_seq', (SELECT MAX(id) FROM wallabag_site_credential));
SELECT setval('tag_id_seq', (SELECT MAX(id) FROM wallabag_tag));
SELECT setval('tagging_rule_id_seq', (SELECT MAX(id) FROM wallabag_tagging_rule));
SELECT setval('user_id_seq', (SELECT MAX(id) FROM wallabag_user));
COMMIT;
```

I tried to negate all changes to the tables and columns themselves, just loading the data and then resetting the starting points of the sequences so new entries can be added correctly. This is the cleanest end result I've managed. If anyone feels like writing a tool from this that adds a UI and makes it easier for newbies, feel free.
